src/TestUtils/src/Microsoft.Maui.IntegrationTests/TemplateTests.cs

```diff
@@ -113,6 +113,62 @@ public void BuildUnpackaged(string id, string framework, string config)
 				$"Project {Path.GetFileName(projectFile)} failed to build. Check test output/attachments for errors.");
 		}
 
+		[Test]
+		[TestCase("maui", $"{DotNetCurrent}-ios", "ios-arm64")]
+		public void PublishNativeAOT(string id, string framework, string runtimeIdentifier)
+		{
+			if (!TestEnvironment.IsMacOS)
+				Assert.Ignore("Publishing a MAUI iOS app with NativeAOT is only supported on a host MacOS system.");
+
+			var projectDir = TestDirectory;
+			var projectFile = Path.Combine(projectDir, $"{Path.GetFileName(projectDir)}.csproj");
+
+			Assert.IsTrue(DotnetInternal.New(id, projectDir, DotNetCurrent),
+				$"Unable to create template {id}. Check test output for errors.");
+
+			var extendedBuildProps = BuildProps;
+			extendedBuildProps.Add("PublishAot=true");
+			extendedBuildProps.Add("PublishAotUsingRuntimePack=true");	// TODO: This parameter will become obsolete https://github.com/dotnet/runtime/issues/87060
+
+			Assert.IsTrue(DotnetInternal.Publish(projectFile, "Release", framework: framework, properties: extendedBuildProps, runtimeIdentifier: runtimeIdentifier),
+				$"Project {Path.GetFileName(projectFile)} failed to build. Check test output/attachments for errors.");
+		}
+
+		[Test]
+		[TestCase("maui", $"{DotNetCurrent}-ios", "ios-arm64")]
+		public void PublishNativeAOTCheckWarnings(string id, string framework, string runtimeIdentifier)
+		{
+			if (!TestEnvironment.IsMacOS)
+				Assert.Ignore("Publishing a MAUI iOS app with NativeAOT is only supported on a host MacOS system.");
+
+			var expectedWarningsCount = 69; // NOTE: As we progress we will want to lower this number to match the current state
```

Comment:
I'm a bit concerned about this seemingly arbitrary warning count that is being tracked, we will presumably end up needing to update this test often and I worry about it adding regular noise to CI. Would it be possible to add a more explicit check for some warning text we _shouldn't_ be seeing? Or maybe we should wait to assert on this until the check can be made more precise?

Reply:
I think the general idea is we have 69 NativeAOT warnings now and they want to make these lower over time. It's going to probably take months to get there.

@ivanpovazan is there a way to create a file that lists the acceptable warnings? As we fix them, we could remove lines from this file.

Reply:
For iOS I asserted on the exact warnings: https://github.com/xamarin/xamarin-macios/commit/638e6f5f1212e771d3f3e5d2727a8a9cf3f679af

Reply:
Thank you all for your feedback.

First of all, I agree with the approach of including more information, than just comparing the number of warnings, for two reasons:
1. It will be a much better/precise verification
2. If a test starts failing, it will force all contributors to become aware of this important behaviour that we are trying to fix

### Proposal

To do this, I propose the following changes:
1. Build a template app with net8.0 release and generate a `.binlog` file
2. Store this `.binlog` as the `expected binlog` result
3. Add a dependency on https://msbuildlog.com/#api to the `IntergrationTests`
4. During IntegrationTest startup, parse the `expected binlog` with the `Microsoft.Build.Logging.StructuredLogger` API and collect `BuildWarningEventArgs` into a `ExpectedWarnings` list
5. Sort the list with some criteria like: by file->by warning code->by warning message (this is needed as the order of warnings is nondeterministic)
6. In the `PublishNativeAOTCheckWarnings` test generate/parse the test `.binlog`, collect `BuildWarningEventArgs`, sort them, and compare against the `ExpectedWarnings`
    - during comparison we will ignore/filter some aspects e.g., ignore root when comparing source file location

### Maintainability  

The concern about updating the test case, as @pjcollins suggested, is there, and we need to understand how frequently we would see a failure, and what we will have to do about it.

I tried to think about the cases (maybe I am missing some) when a test might fail, and I tried to order them by severity:
1) Moving code around where a code that produced some warning is now placed in a different (or renamed) file - will fail to match the expected source file location
2) Renaming types/members that are reporting the warning - will fail to match the expected warning message
3) Code changes that fix warnings - will fail for missing expected warnings
4) Code changes that introduce new warnings - will fail because of new warnings
5) Code changes that fix warnings but introduce new/different ones - will fail because of missing expected and new warnings

Fixing any of the above comes down to: 
- Generate a new `.binlog` with the current state which would become the new `expected binlog` and store/commit that as the new default. 

**IMPORTANT NOTE:**  However, it is very important to understand why and when the update of the expected result has to be done, especially in cases of 4) and 5) where a change is essentially working against our efforts to reach 0 trim/AOT warnings with MAUI, and will have to be evaluated and addressed sooner rather than later.

---
@rolfbjarne @jonathanpeppers @pjcollins What do you think?

PS I also considered storing a simple list of all the 69 warning messages strings to be the `expected value`, but that didn't seem like a viable option because in case of a failure updating such list would become problematic for any/all of the reasons a failure can occur as described above.

Reply:
/cc: @akoeplinger thanks for the suggestion on using the msbuildlog API, please see my proposal above

Reply:
I don't like storing the binlog, a flat text file of the extracted warnings strings sounds much better to me, I don't really understand what you mean with "in case of a failure updating such list would become problematic for any/all of the reasons a failure can occur as described above." ?

Reply:
I agree with @akoeplinger, a list of warnings in code/text is much smaller and easier to reason about than a binary file.

It also makes diffs easier to read, it's rather obvious here that the dotnet/runtime bump fixed a warning we were getting, no explanation needed: https://github.com/xamarin/xamarin-macios/commit/73e15f09babc136aae9a6eb15826b0a61fce3b10 - this would show up very differently if the binlog changed.

Reply:
Maybe I am overthinking, but if we just store the warnings as a flat text and compare against it, I fear that the developer experience wont be really good. Consider the following example:
- Lets say there are 4 mismatches on lines: 27, 39, 47 and 49.
To what assembly/file do these lines correspond to, are the failures related to renaming a type, should the changes in the PR with these failures be corrected, or should the expected output be updated - and if it does, how do we update this flat text (it is not just copy-pasting from the build output as the output includes source file and line number information)? 

If we would have just a few warnings, we probably could easily store/compare against them, but with 69 warnings I think it will quickly become cumbersome to maintain this flat text file.

Reply:
I would ignore the line number, and group by file, then warning number/text.

So you'd assert that in the file myfile.cs there are 2 warnings of IL0001, and 3 warnings of IL0002.

If you fix IL0002 in that file, then you remove those asserts.

if you find 4 IL0002 warnings in myfile.cs, you assert with a message "Expected 3 IL0002 warnings in myfile.cs, got 4"

If code is moved around between files, the test would have to be updated accordingly.

Code-wise, it would be something like this (pseudo-code, I don't remember the API exactly):

```cs
class WarningInfo {
	public int Count;
    public string Code;
    public string Message;
}
var actualWarnings = GetAllWarningsFromBinlog ();
var expectedWarnings = new [] {
	new { File = "file1.cs", Warnings = new [
		new WarningInfo { Count = 2, Code = "IL0001", Message = "..." },
		new WarningInfo { Count = 3, Code = "IL0002", Message = "..." }
        ] },
    new { File = "file2.cs", Warnings = new [ ... ] },
};

// Assert that we get warnings in all files we expect to, and none if files we don't
var actualFiles = actualWarnings.Select (w => w.File).Distinct ();
var expectedFiles = expectedWarnings.Slect (w => w.File);
CollectionAssert.AreEquivalent (expectedFiles, actualFiles, "Files with warnings");

// For each file, check that we got the number of expected warnings for each warning code.
foreach (var file in expectedWarnings) {
    var actualFile = actualWarnings.Where (v => v.File == file.File).ToList ();
    foreach (var info in file.Warnings) {
        var actualPerCode = actualFile.Where (v => v.Code == info.Code);
        // assert that we got the expected number of warnings for the specified warning code
        Assert.AreEqual (info.Warnings.Length, actualPerCode.Count (), $"Warning {info.Code} in {info.File}");
        actualFile.Remove (v => v.Code == info.Code);
    }
    // assert that we didn't get any warnings we didn't expect
    Assert.AreEqual (0, actualFile.Count, $"Unexpeced warnings in {file.File}:\n\t{string.Join ("\n\t\t", actualFile.Select (v => v.ToString ()))}");
}

// If I got the logic above correct, there shouldn't be a need to assert the total number of warnings, but just in case.
// Also do this last, so that we get more detailed asserts from above if those trigger.
var expectedTotalWarningCount = expectedWarnings.<SomeLinqExpressionHereToSumAllTheCountValues> ();
Assert.AreEqual (expectedTotalWarningCount, actualWarnings.Count (), "Total number of warnings");
```

Reply:
Thank you for the detailed explanation.

I have two questions regarding the `Message` it self (you might have left out the logic in the pseudo-code for simplicity):
1. From the provided approach it seems we wouldn't actually check the content of the message, but rather just they type of the warning and its count per file. Do we want to check the message content?
2. If we do want to check for the exact `Message` wouldn't we need to have another dimension in the `WarningInfo` type?

Reply:
Yes, I would assert on the Message too, I skipped that to make the code simpler.

As you say, adding another dimension should work to accomplish this I believe.

Making the Message property a string array could also work.

Reply:
I implemented @rolfbjarne's approach, so please take a look at the new implementation.
I will mark the conversation as resolved for clarity.